tsdb/engine/tsm1/file_store.go

```diff
@@ -782,12 +840,25 @@ func newKeyCursor(fs *FileStore, key string, t int64, ascending bool) *KeyCursor
 		sort.Sort(descLocations(c.seeks))
 	}
 
+	// Determine the distinct set of TSM files in use and mark then as in-use
+	for _, f := range c.seeks {
+		if _, ok := c.refs[f.r.Path()]; !ok {
+			f.r.Ref()
```

Comment:
There isn't a race.  There is a read-lock taken on the file store by the caller of `newKeyCursor`.  The cursor then uses the TSM files at that point in time until they complete.  The file store is free to remove the references to the `TSMFile`s it's managing when it's able to acquire the write lock.